staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go

```diff
@@ -31,171 +31,278 @@ import (
+// Returns false if value is not found or is not a []interface{}.
+func NestedStringSlice(obj map[string]interface{}, fields ...string) ([]string, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.([]interface{}); ok {
 		strSlice := make([]string, 0, len(m))
 		for _, v := range m {
 			if str, ok := v.(string); ok {
 				strSlice = append(strSlice, str)
+			} else {
+				return nil, false
 			}
 		}
-		return strSlice
+		return strSlice, true
 	}
-	return nil
+	return nil, false
+}
+
+// NestedSlice returns the []interface{} value of a nested field.
+// Returns false if value is not found or is not a []interface{}.
+func NestedSlice(obj map[string]interface{}, fields ...string) ([]interface{}, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.([]interface{}); ok {
+		slice := make([]interface{}, len(m))
+		copy(slice, m)
+		return slice, true
+	}
+	return nil, false
 }
 
-func getNestedMap(obj map[string]interface{}, fields ...string) map[string]string {
-	if m, ok := getNestedField(obj, fields...).(map[string]interface{}); ok {
+// NestedStringMap returns the map[string]string value of a nested field.
+// Returns false if value is not found or is not a map[string]interface{}.
+func NestedStringMap(obj map[string]interface{}, fields ...string) (map[string]string, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.(map[string]interface{}); ok {
 		strMap := make(map[string]string, len(m))
 		for k, v := range m {
 			if str, ok := v.(string); ok {
 				strMap[k] = str
+			} else {
+				return nil, false
 			}
 		}
-		return strMap
+		return strMap, true
+	}
+	return nil, false
+}
+
+// NestedMap returns the map[string]interface{} value of a nested field.
+// Returns false if value is not found or is not a map[string]interface{}.
+func NestedMap(obj map[string]interface{}, fields ...string) (map[string]interface{}, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.(map[string]interface{}); ok {
+		nMap := make(map[string]interface{}, len(m))
+		for k, v := range m {
+			nMap[k] = v
+		}
+		return nMap, true
+	}
+	return nil, false
+}
+
+func assertValueType(val interface{}) {
+	switch x := val.(type) {
+	case map[string]interface{}:
+		for _, v := range x {
+			assertValueType(v)
+		}
+	case []interface{}:
+		for _, v := range x {
+			assertValueType(v)
+		}
+	case string, int64, bool, float64, nil, gojson.Number:
+	default:
+		panic(fmt.Errorf("invalid value type: %T", val))
 	}
-	return nil
 }
 
-func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+// SetNestedField sets the value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedField(obj map[string]interface{}, value interface{}, fields ...string) bool {
```

Comment:
I wished we had a debugging/strict-test mode where `value` is type checked. We use(d) to write arbitrary values into objects (https://github.com/kubernetes/kubernetes/pull/54803) and the compiler cannot help us with type checks.

Reply:
I can send a follow up PR with a environment variable flag a-la https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go#L81 and https://github.com/kubernetes/kubernetes/blob/ed423054ba0089a6d58fb0e048fe1acfd966f0d7/staging/src/k8s.io/client-go/tools/cache/mutation_detector.go#L34

Reply:
Would be great to run unit and maybe even integration tests with this.

Reply:
Follow up in https://github.com/kubernetes/kubernetes/pull/55297